src/java/org/apache/cassandra/utils/ByteSource.java

```diff
@@ -0,0 +1,775 @@
+     * E.g. A00B translates to 4100FEFF4200
+     *      A0B0               4100FF4200FE (+00 for OSS41)
+     *      A0                 4100FE       (+00 for OSS41)
+     *
+     * If in a single byte source, the bytes could be simply passed unchanged, but this would not allow us to
+     * combine components. This translation preserves order, and since the encoding for 0 is higher than the separator
+     * also makes sure shorter components are treated as smaller.
+     *
+     * The encoding is not prefix-free, since e.g. the encoding of "A" (4100) is a prefix of the encoding of "A0"
+     * (4100FE), but the byte following the prefix is guaranteed to be FE or FF, which makes the encoding weakly
+     * prefix-free. Additionally, any such prefix sequence will compare smaller than the value to which it is a prefix,
+     * because any permitted separator byte will be smaller than the byte following the prefix.
+     *
+     * See also : https://docs.google.com/document/d/1X9S69gh2-qkb5DyWOFtu-DAr4tBKtzmGJP-Glr8uKL8/edit
+     */
+    static abstract class AbstractReinterpreter implements ByteSource
+    {
+        final ByteComparable.Version version;
+        int bufpos;
+        boolean escaped;
+
+        AbstractReinterpreter(int position, ByteComparable.Version version)
+        {
+            this.bufpos = position;
+            this.version = version;
+        }
+
+        public final int next()
+        {
+            if (bufpos >= limit())
+            {
+                if (bufpos > limit())
+                    return END_OF_STREAM;
+
+                ++bufpos;
+                if (escaped)
+                {
+                    escaped = false;
+                    --bufpos;
+                    return ESCAPED_0_CONT;
+                }
+                return ESCAPE;
+            }
+
+            int index = bufpos++;
+            int b = get(index) & 0xFF;
+            if (!escaped)
+            {
+                if (b == ESCAPE)
+                    escaped = true;
+                return b;
+            }
+            else
+            {
+                if (b == ESCAPE)
+                    return ESCAPED_0_CONT;
+                --bufpos;
+                escaped = false;
+                return ESCAPED_0_DONE;
+            }
+        }
+
+        protected abstract byte get(int index);
+
+        protected abstract int limit();
+    }
+
+    static class BufferReinterpreter extends AbstractReinterpreter
+    {
+        final ByteBuffer buf;
+
+        BufferReinterpreter(ByteBuffer buf, ByteComparable.Version version)
+        {
+            super(buf.position(), version);
+            this.buf = buf;
+            this.bufpos = buf.position();
+        }
+
+        protected int limit()
+        {
+            return buf.limit();
+        }
+
+        protected byte get(int index)
+        {
+            return buf.get(index);
+        }
+    }
+
+    static class UnsafeReinterpreter extends AbstractReinterpreter
```

Comment:
We are not using this at the moment, it may be better to leave it with all the `Unsafe` infra it brings out for now.

Reply:
I may extract that stuff into separate commit if needed